File: sdk/golang/internal/infra/default_ws_client.go
package infra

import (
	"context"
	"fmt"
	"math/rand"
	"net/url"
	"sync"
	"sync/atomic"
	"time"

	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/internal/interfaces"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/common/logger"
	"github.com/Kucoin/kucoin-universal-sdk/sdk/golang/pkg/types"
	"github.com/gorilla/websocket"
)

func NewPingMessage() *types.WsMessage {
	return &types.WsMessage{
		Id:   IntToString(time.Now().UnixNano()),
		Type: types.PingMessage,
	}
}

type writeMsg struct {
	signal chan error
	ctx    context.Context
	msg    *types.WsMessage
	time   int64
}

// WebSocketClient represents the WebSocket client
type WebSocketClient struct {
	options          *types.WebSocketClientOption
	conn             *websocket.Conn
	connLock         sync.Mutex
	connected        atomic.Bool
	shutdown         atomic.Bool
	disconnectEvent  chan struct{}
	reconnectedEvent chan struct{}

	tokenProvider      interfaces.WsTokenProvider
	tokenInfo          *interfaces.WsToken
	closeChan          chan struct{}
	reconnectCloseChan chan struct{}

	writeMsg chan *writeMsg
	readMsg  chan *types.WsMessage

	ackEvent     map[string]*writeMsg
	ackEventLock sync.Mutex

	wg     sync.WaitGroup
	metric struct {
		pingSuccess int64
		pingErr     int64
		goroutines  int64
	}
}

// NewWebSocketClient creates a new WebSocketClient instance
func NewWebSocketClient(tokenProvider interfaces.WsTokenProvider, options *types.WebSocketClientOption) *WebSocketClient {
	return &WebSocketClient{
		options:            options,
		conn:               nil,
		connLock:           sync.Mutex{},
		connected:          atomic.Bool{},
		shutdown:           atomic.Bool{},
		disconnectEvent:    make(chan struct{}, 1),
		reconnectedEvent:   make(chan struct{}, 1),
		tokenProvider:      tokenProvider,
		tokenInfo:          nil,
		closeChan:          make(chan struct{}),
		reconnectCloseChan: make(chan struct{}),
		writeMsg:           make(chan *writeMsg, options.WriteMessageBuffer),
		readMsg:            make(chan *types.WsMessage, options.ReadMessageBuffer),
		ackEvent:           make(map[string]*writeMsg),
		ackEventLock:       sync.Mutex{},
		wg:                 sync.WaitGroup{},
	}
}

// Start establishes the WebSocket connection
func (c *WebSocketClient) Start() error {
	c.connLock.Lock()
	defer c.connLock.Unlock()

	if c.connected.Load() {
		return fmt.Errorf("already connected")
	}

	err := c.dial()
	if err != nil {
		return err
	}
	c.connected.Store(true)
	c.notifyEvent(types.EventConnected, "")
	c.run()
	c.reconnect()
	return nil
}

// Stop closes the WebSocket connection
func (c *WebSocketClient) Stop() error {
	c.connLock.Lock()
	defer c.connLock.Unlock()
	c.notifyEvent(types.EventClientShutdown, "")
	c.shutdown.Store(true)
	c.close()
	return nil
}

func (c *WebSocketClient) Reconnected() <-chan struct{} {
	return c.reconnectedEvent
}

func (c *WebSocketClient) notifyEvent(event types.WebSocketEvent, msg string) {
	defer func() {
		if r := recover(); r != nil {
			logger.GetLogger().Errorf("recovered in notifyEvent: %v", r)
		}
	}()

	if c.options.EventCallback != nil {
		c.options.EventCallback(event, msg)
	}
}

func (c *WebSocketClient) run() {
	c.wg.Add(3)
	atomic.AddInt64(&c.metric.goroutines, 3)
	go func() {
		defer func() {
			atomic.AddInt64(&c.metric.goroutines, -1)
			c.wg.Done()
		}()
		c.keepAlive()
	}()
	go func() {
		defer func() {
			atomic.AddInt64(&c.metric.goroutines, -1)
			c.wg.Done()
		}()
		c.readMessages()
	}()
	go func() {
		defer func() {
			atomic.AddInt64(&c.metric.goroutines, -1)
			c.wg.Done()
		}()
		c.writeMessage()
	}()
}

func (c *WebSocketClient) Write(ctx context.Context, ms *types.WsMessage) <-chan error {
	if !c.connected.Load() {
		ch := make(chan error, 1)
		ch <- fmt.Errorf("not connected")
		return ch
	}

	c.ackEventLock.Lock()
	defer c.ackEventLock.Unlock()

	msg := &writeMsg{
		signal: make(chan error, 1),
		ctx:    ctx,
		msg:    ms,
		time:   time.Now().Unix(),
	}

	c.ackEvent[ms.Id] = msg
	c.writeMsg <- msg
	return msg.signal
}

func (c *WebSocketClient) Read() <-chan *types.WsMessage {
	return c.readMsg
}

// dial establishes the WebSocket connection
func (c *WebSocketClient) dial() (err error) {
	defer func() {
		if err == nil {
			logger.GetLogger().Info("connection established")
		} else {
			logger.GetLogger().Errorf("failed to connect: %v", err)
		}
	}()

	err, tokenInfo := c.tokenProvider.GetToken()
	if err != nil {
		return fmt.Errorf("get token error, %s", err.Error())
	}

	token, err := c.randomEndpoint(tokenInfo)
	if err != nil {
		return err
	}

	c.tokenInfo = token

	// dail websocket
	dialer := websocket.Dialer{
		HandshakeTimeout: c.options.DialTimeout,
		ReadBufferSize:   c.options.ReadBufferBytes,
	}
	q := url.Values{}
	q.Add("connectId", IntToString(time.Now().UnixNano()))
	q.Add("token", token.Token)
	urlStr := fmt.Sprintf("%s?%s", token.Endpoint, q.Encode())

	conn, _, err := dialer.Dial(urlStr, nil)
	if err != nil {
		return err
	}
	c.conn = conn

	// check welcome message
	m := &types.WsMessage{}
	if err := c.conn.ReadJSON(m); err != nil {
		return err
	}

	if m.Type == types.ErrorMessage {
		return fmt.Errorf(string(m.RawData))
	}

	if m.Type != types.WelcomeMessage {
		return fmt.Errorf("not receive welcome message")
	}

	return nil
}

// readMessages continuously reads messages from the WebSocket connection
func (c *WebSocketClient) readMessages() {
	for {

		select {
		case <-c.closeChan:
			{
				return
			}
		default:
		}

		m := &types.WsMessage{}

		if err := c.conn.ReadJSON(m); err != nil {
			logger.GetLogger().Errorf("websocket connection got error: %v", err)
			if !websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
				c.disconnectEvent <- struct{}{}
				return
			}
			continue
		}

		switch m.Type {
		case types.Message:
			{
				c.notifyEvent(types.EventMessageReceived, "")
				select {
				case c.readMsg <- m:
					{
						// ok
					}
				default:
					{
						//warn and break
						c.notifyEvent(types.EventReadBufferFull, "")
						logger.GetLogger().Errorf("read buffer full")
					}
				}
				break
			}

		case types.PongMessage:
			c.notifyEvent(types.EventPongReceived, "")
			fallthrough
		case types.AckMessage:
			fallthrough
		case types.ErrorMessage:
			{
				func() {
					c.ackEventLock.Lock()
					defer c.ackEventLock.Unlock()

					msg, exist := c.ackEvent[m.Id]
					if !exist {
						logger.GetLogger().Errorf("websocket can not find ack event, id:%s", m.Id)
						return
					}
					var err error
					if m.Type == types.ErrorMessage {
						err = fmt.Errorf(string(m.RawData))
						c.notifyEvent(types.EventErrorReceived, err.Error())
					}
					msg.signal <- err
					delete(c.ackEvent, m.Id)
				}()
			}
		default:
			{
				logger.GetLogger().Errorf("websocket unknown type: %v", m.Type)
			}
		}
	}
}

func (c *WebSocketClient) writeMessage() {
	for {
		select {
		case data := <-c.writeMsg:
			select {
			case <-data.ctx.Done():
				// timeout
				logger.GetLogger().Warnf("websocket wrtie data already exceed deadline, id: %v", data.msg.Id)
				data.signal <- data.ctx.Err()
				c.ackEventLock.Lock()
				delete(c.ackEvent, data.msg.Id)
				c.ackEventLock.Unlock()
				continue
			default:
				// pass
			}

			c.conn.SetWriteDeadline(time.Now().Add(c.options.WriteTimeout))
			err := c.conn.WriteJSON(data.msg)
			if err != nil {
				// remove ack
				data.signal <- err
				c.ackEventLock.Lock()
				delete(c.ackEvent, data.msg.Id)
				c.ackEventLock.Unlock()
				logger.GetLogger().Errorf("websocket write err: %v", err)
			}
		case <-c.closeChan:
			return
		}
	}
}

// keepAlive sends periodic ping messages to keep the connection alive
func (c *WebSocketClient) keepAlive() {
	ticker := time.NewTicker(time.Duration(c.tokenInfo.PingInterval) * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			{
				func() {
					ctx, cancel := context.WithTimeout(context.Background(), time.Duration(c.tokenInfo.PingTimeout)*time.Millisecond)
					defer cancel()

					ping := NewPingMessage()
					writeSignal := c.Write(ctx, ping)
					select {
					case err := <-writeSignal:
						{
							if err == nil {
								c.metric.pingSuccess++
								logger.GetLogger().Debugf("heartbeat ping ok")
							} else {
								logger.GetLogger().Errorf("heartbeat ping error: %v", err)
								c.metric.pingErr++
							}
						}
					case <-ctx.Done():
						logger.GetLogger().Errorf("heartbeat ping timeout")
						c.metric.pingErr++
					}
				}()
			}
		case <-c.closeChan:
			{
				return
			}
		}
	}
}

func (c *WebSocketClient) reconnect() {
	go func() {
		for {
			select {
			case <-c.reconnectCloseChan:
				return

			case <-c.disconnectEvent:

				if c.shutdown.Load() {
					continue
				}

				logger.GetLogger().Infof("broken websocket connection, start reconnect")

				c.close()
				c.notifyEvent(types.EventTryReconnect, "")

				attempt := 0
				reconnected := false
				for {
					// Handle reconnect attempts
					if !c.options.Reconnect || (c.options.ReconnectAttempts != -1 && attempt >= c.options.ReconnectAttempts) {
						logger.GetLogger().Errorf("max reconnect attempts reached or reconnect disabled")
						break
					}

					logger.GetLogger().Infof("reconnecting in %v... (attempt %d)", c.options.ReconnectInterval, attempt)
					time.Sleep(c.options.ReconnectInterval)

					err := c.dial()
					if err == nil {
						c.notifyEvent(types.EventConnected, "")
						c.closeChan = make(chan struct{})
						c.connected.Store(true)
						c.run()
						c.reconnectedEvent <- struct{}{}
						reconnected = true
						break
					}

					attempt++
				}

				if reconnected {
					continue
				}

				c.notifyEvent(types.EventClientFail, "")
			}
		}
	}()
}

func (c *WebSocketClient) randomEndpoint(tokens []*interfaces.WsToken) (*interfaces.WsToken, error) {
	if tokens == nil || len(tokens) == 0 {
		return nil, fmt.Errorf("tokens is empty")
	}
	return tokens[rand.Intn(len(tokens))], nil
}

func (c *WebSocketClient) close() {
	if c.connected.CompareAndSwap(true, false) {
		close(c.closeChan)
		c.conn.Close()
		c.wg.Wait()
		c.conn = nil
		c.notifyEvent(types.EventDisconnected, "")

		c.ackEventLock.Lock()
		// clean up events
		for _, event := range c.ackEvent {
			select {
			case event.signal <- fmt.Errorf("connection closed"):
				// ok
			default:
				// pass
			}
		}
		c.ackEvent = make(map[string]*writeMsg)
		c.ackEventLock.Unlock()

		if c.shutdown.Load() {
			close(c.readMsg)
			close(c.reconnectCloseChan)
		}
	}
	c.tokenProvider.Close()
	logger.GetLogger().Infof("close websocket client")
}

File: sdk/python/kucoin_universal_sdk/internal/infra/default_ws_client.py
import logging
import random
import threading
import time
import urllib.parse
from queue import Queue, Full, Empty
from typing import Dict, List, Optional

import websocket

from kucoin_universal_sdk.model.common import WsMessage
from kucoin_universal_sdk.model.constants import WsMessageType
from kucoin_universal_sdk.model.websocket_option import WebSocketClientOption
from kucoin_universal_sdk.model.websocket_option import WebSocketEvent
from ..interfaces.websocket import WsTokenProvider, WsToken

class WriteMsg:
    def __init__(self, msg: WsMessage, timeout: float):
        self.msg = msg
        self.ts = time.time()
        self.timeout = timeout
        self.exception = None
        self.event = threading.Event()

    def set_exception(self, exception: Exception):
        self.exception = exception
        self.event.set()

class WebSocketClient:
    def __init__(self, token_provider: WsTokenProvider, options: WebSocketClientOption):
        self.options = options
        self.conn = None
        self.conn_lock = threading.Lock()
        self.connected = threading.Event()
        self.shutdown = threading.Event()
        self.disconnect_event = threading.Event()
        self.reconnected_event = threading.Event()
        self.token_provider = token_provider
        self.token_info = None
        self.close_event = threading.Event()
        self.reconnect_close_event = threading.Event()

        self.read_msg = Queue(maxsize=options.read_message_buffer)
        self.write_msg = Queue(maxsize=options.write_message_buffer)

        self.ack_event: Dict[str, WriteMsg] = {}
        self.ack_event_lock = threading.Lock()
        self.metric = {'ping_success': 0, 'ping_err': 0}
        self.keep_alive_thread = None
        self.write_thread = None
        self.ws_thread = None
        self.reconnect_thread = None
        self.welcome_received = threading.Event()

    def start(self):
        with self.conn_lock:
            if self.connected.is_set():
                logging.warning("WebSocket client is already connected.")
                return
            try:
                self.dial()
            except Exception as err:
                logging.error(f"Failed to start WebSocket client: {err}")
                raise
            self.connected.set()
            self.notify_event(WebSocketEvent.EVENT_CONNECTED, "")
            self.run()
            self.reconnect()

    def run(self):
        if not self.keep_alive_thread or not self.keep_alive_thread.is_alive():
            self.keep_alive_thread = threading.Thread(target=self.keep_alive, daemon=True)
            self.keep_alive_thread.start()

        if not self.write_thread or not self.write_thread.is_alive():
            self.write_thread = threading.Thread(target=self.write_message, daemon=True)
            self.write_thread.start()

    def stop(self):
        with self.conn_lock:
            self.notify_event(WebSocketEvent.EVENT_CLIENT_SHUTDOWN, "")
            self.shutdown.set()
            self.close()

    def dial(self):
        try:
            token_infos = self.token_provider.get_token()
            self.token_info = self.random_endpoint(token_infos)
            query_params = {
                "connectId": str(int(time.time() * 1e9)),
                "token": self.token_info.token,
            }
            url_str = f"{self.token_info.endpoint}?{urllib.parse.urlencode(query_params)}"
            self.conn = websocket.WebSocketApp(
                url_str,
                on_message=self.on_message,
                on_error=self.on_error,
                on_close=self.on_close,
                on_open=self.on_open,
            )
            if not self.ws_thread or not self.ws_thread.is_alive():
                self.ws_thread = threading.Thread(target=self.conn.run_forever, daemon=True)
                self.ws_thread.start()
            if not self.welcome_received.wait(timeout=5):
                self.close()
                self.disconnect_event.set()
                raise Exception("Did not receive welcome message")
        except Exception as err:
            self.connected.clear()
            logging.error(f"Failed to connect or validate welcome message: {err}")
            raise

    def on_open(self, ws):
        logging.info("WebSocket connection opened.")

    def on_message(self, ws, message):
        logging.debug(f"Received message: {message}")
        m = WsMessage.from_json(message)
        if m.type == WsMessageType.WELCOME.value:
            self.welcome_received.set()
            logging.info("Welcome message received.")

        elif m.type == WsMessageType.MESSAGE.value:
            self.notify_event(WebSocketEvent.EVENT_MESSAGE_RECEIVED, "")
            try:
                logging.debug(f"queue size: {self.read_msg.qsize()}, max size: {self.read_msg.maxsize}")
                self.read_msg.put(m, block=False)
            except Full:
                self.notify_event(WebSocketEvent.EVENT_READ_BUFFER_FULL, "")
                logging.warning("Read buffer full")

        elif m.type == WsMessageType.PONG.value:
            self.notify_event(WebSocketEvent.EVENT_PONG_RECEIVED, "")
            logging.debug("PONG received")
            self._handle_ack_event(m)

        elif m.type in [WsMessageType.ACK.value, WsMessageType.ERROR.value]:
            self._handle_ack_event(m)

        else:
            logging.warning(f"Unknown message type: {m.type}")

    def _handle_ack_event(self, m: WsMessage):
        with self.ack_event_lock:
            data: WriteMsg = self.ack_event.pop(m.id, None)
            if not data:
                logging.warning(f"Cannot find ack event, id: {m.id}")
                return
            if m.type == WsMessageType.ERROR.value:
                error = m.raw_data
                self.notify_event(WebSocketEvent.EVENT_ERROR_RECEIVED, error)
                data.set_exception(error)
            else:
                data.event.set()

    def read(self):
        return self.read_msg

    def write(self, ms: WsMessage, timeout: float) -> WriteMsg:
        logging.info(f"Write message: {ms}")
        if not self.connected.is_set():
            raise Exception("Not connected")

        msg = WriteMsg(msg=ms, timeout=timeout)
        with self.ack_event_lock:
            self.ack_event[ms.id] = msg
            try:
                self.write_msg.put(msg)
            except Full:
                logging.warning(f"Write buffer is full for message ID {ms.id}.")
                self.ack_event.pop(ms.id, None)
            except Exception as e:
                logging.error(f"Exception in write method: {e}")
                self.ack_event.pop(ms.id, None)
                raise
            finally:
                return msg

    def write_message(self):
        while not self.close_event.is_set():
            try:
                data: WriteMsg = self.write_msg.get(timeout=1)
                if self.conn:
                    self.conn.send(data.msg.to_json())
                    logging.debug(f"Message sent: {data.msg}")
                    data.ts = time.time()
                else:
                    logging.warning("No connection available to send message.")
            except Empty:
                continue
            except Exception as e:
                logging.error(f"Error sending message: {e}")

    def keep_alive(self):
        interval = self.token_info.ping_interval / 1000.0
        timeout = self.token_info.ping_timeout / 1000.0
        last_ping_time = time.time()
        
        while not self.shutdown.is_set() and not self.close_event.is_set():
            current_time = time.time()
            if current_time - last_ping_time >= interval:
                ping_msg = self.new_ping_message()
                try:
                    self.write(ping_msg, timeout=timeout)
                    self.metric['ping_success'] += 1
                except TimeoutError:
                    logging.error("Heartbeat ping timeout")
                    self.metric['ping_err'] += 1
                except Exception as e:
                    logging.error(f"Exception in keep_alive: {e}")
                    self.metric['ping_err'] += 1
                last_ping_time = current_time
            
            time.sleep(1)

    def on_error(self, ws, error):
        logging.error(f"WebSocket error: {error}")
        self.disconnect_event.set()

    def on_close(self, ws, close_status_code, close_msg):
        self.connected.clear()
        self.disconnect_event.set()
        logging.info(f"WebSocket closed with status code {close_status_code}, message: {close_msg}")

    def reconnect(self):
        def reconnect_loop():
            while True:
                if self.reconnect_close_event.wait(timeout=1):
                    return

                if self.disconnect_event.wait(timeout=1):
                    if self.shutdown.is_set():
                        continue

                    logging.info("Broken WebSocket connection, starting reconnection")
                    self.close()
                    self.notify_event(WebSocketEvent.EVENT_TRY_RECONNECT, "")
                    self.disconnect_event.clear()

                    attempt = 0
                    reconnected = False
                    if not self.options.reconnect or (
                        self.options.reconnect_attempts != -1 and attempt >= self.options.reconnect_attempts):
                        logging.error("Max reconnect attempts reached or reconnect disabled")
                        break

                    logging.info(
                        f"Reconnecting in {self.options.reconnect_interval} seconds... (attempt {attempt})")
                    time.sleep(self.options.reconnect_interval)

                    try:
                        self.dial()
                        self.notify_event(WebSocketEvent.EVENT_CONNECTED, "")
                        self.connected.set()
                        self.run()
                        reconnected = True
                        break
                    except Exception as err:
                        logging.error(f"Reconnect attempt {attempt} failed: {err}")
                        attempt += 1

                    if reconnected:
                        self.reconnected_event.set()
                        continue

                    self.notify_event(WebSocketEvent.EVENT_CLIENT_FAIL, "")
                    logging.error("Failed to reconnect after all attempts.")

        self.reconnect_thread = threading.Thread(target=reconnect_loop)
        self.reconnect_thread.start()

    def _clear_message_queues(self):
        while not self.read_msg.empty():
            self.read_msg.get_nowait()
        while not self.write_msg.empty():
            self.write_msg.get_nowait()

    def close(self):
        if self.connected.is_set():
            self.shutdown.set()
            self.disconnect_event.set()
            self.reconnect_close_event.set()
            self.connected.clear()

            with self.ack_event_lock:
                for msg in self.ack_event.values():
                    msg.event.set()
                self.ack_event.clear()
            self._clear_message_queues()

            if self.conn:
                self.conn.close()
                self.conn = None
                self.close_event.set()
                logging.info("WebSocket connection closed.")
        logging.info("Waiting all threads close...")
        self.token_provider.close()
        self.write_thread.join()
        self.keep_alive_thread.join()
        self.ws_thread.join()
        self.reconnect_thread.join()
        self.notify_event(WebSocketEvent.EVENT_DISCONNECTED, "")
        logging.info("WebSocket client closed.")

    def notify_event(self, event: WebSocketEvent, msg: str, msg2=""):
        try:
            if self.options.event_callback is not None:
                self.options.event_callback(event, msg, msg2)
        except Exception as e:
            logging.error(f"Exception in notify_event: {e}")

    def random_endpoint(self, tokens: List[WsToken]) -> Optional[WsToken]:
        if not tokens:
            raise ValueError("Tokens list is empty")
        return random.choice(tokens)

    def new_ping_message(self) -> WsMessage:
        return WsMessage(
            id=str(int(time.time() * 1e9)),
            type=WsMessageType.PING.value,
        )
